statsd/uds.go

```diff
@@ -20,33 +25,90 @@ type udsWriter struct {
 }
 
 // newUDSWriter returns a pointer to a new udsWriter given a socket file path as addr.
-func newUDSWriter(addr string, writeTimeout time.Duration) (*udsWriter, error) {
-	udsAddr, err := net.ResolveUnixAddr("unixgram", addr)
-	if err != nil {
-		return nil, err
-	}
+func newUDSWriter(addr string, writeTimeout time.Duration, transport string) (*udsWriter, error) {
 	// Defer connection to first Write
-	writer := &udsWriter{addr: udsAddr, conn: nil, writeTimeout: writeTimeout}
+	writer := &udsWriter{addr: addr, transport: transport, conn: nil, writeTimeout: writeTimeout}
 	return writer, nil
 }
 
+// retryOnWriteErr returns true if we should retry writing after a write error
+func (w *udsWriter) retryOnWriteErr(err error) bool {
+	if err == nil {
+		return true
+	}
+	// Never retry when using unixgram (to preserve the historical behavior)
+	if w.transport == "unixgram" {
+		return false
+	}
+	// Otherwise we retry on timeout because we might have written a partial packet
+	if networkError, ok := err.(net.Error); ok && networkError.Timeout() {
+		return true
+	}
+	return false
+}
+
+func (w *udsWriter) shouldCloseConnection(err error) bool {
+	if err, isNetworkErr := err.(net.Error); err != nil && (!isNetworkErr || !err.Timeout()) {
+		// Statsd server disconnected, retry connecting at next packet
+		return true
+	}
+	return false
+}
+
+// writeFull writes the whole data to the UDS connection
+func (w *udsWriter) writeFull(data []byte, stopIfNoneWritten bool) (int, error) {
+	written := 0
+	for written < len(data) {
+		n, e := w.conn.Write(data)
+		written += n
+
+		// If we haven't written anything and we're supposed to stop if we can't write anything, return the error
+		if written == 0 && stopIfNoneWritten {
+			return written, e
+		}
+
+		// If there's an error, check if it is retryable
+		if e != nil && !w.retryOnWriteErr(e) {
+			return written, e
+		}
+
+		// When using "unix" we need to be able to finish to write partially written packets once we have started.
+		if w.transport == "unix" {
```

Comment:
Confidence check:
```
	w.transport = newConn.RemoteAddr().Network()
```
This always returns `unix`, never `unixstream` or similar? Otherwise this current line should be adapted.

Reply:
```
// Network returns the address's network name, "unix", "unixgram" or
// "unixpacket".
func (a *UnixAddr) Network() string {
	return a.Net
}
```

I think we're good ?